managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -1033,26 +1041,81 @@ public void asyncReadEntriesWithSkipOrWait(int maxEntries, long maxSizeBytes, Re
-                return;
-            }
+            synchronized (pendingReadOpMutex) {
+                if (!WAITING_READ_OP_UPDATER.compareAndSet(this, null, op)) {
+                    op.recycle();
+                    callback.readEntriesFailed(new ManagedLedgerException.ConcurrentWaitCallbackException(), ctx);
+                    return;
+                }
 
-            if (log.isDebugEnabled()) {
-                log.debug("[{}] [{}] Deferring retry of read at position {}", ledger.getName(), name, op.readPosition);
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] [{}] Deferring retry of read at position {}", ledger.getName(), name,
+                            op.readPosition);
+                }
+
+                // Check again for new entries after the configured time, then if still no entries are available register
+                // to be notified
+                if (getConfig().getNewEntriesCheckDelayInMillis() > 0) {
+                    delayCheckForNewEntriesTask = new DelayCheckForNewEntriesTask(op, callback, ctx);
+                } else {
+                    // If there's no delay, check directly from the same thread
+                    checkForNewEntries(op, callback, ctx);
+                }
             }
+        }
+    }
 
-            // Check again for new entries after the configured time, then if still no entries are available register
-            // to be notified
-            if (getConfig().getNewEntriesCheckDelayInMillis() > 0) {
-                ledger.getScheduledExecutor()
-                        .schedule(() -> checkForNewEntries(op, callback, ctx),
-                                getConfig().getNewEntriesCheckDelayInMillis(), TimeUnit.MILLISECONDS);
-            } else {
-                // If there's no delay, check directly from the same thread
+    private class DelayCheckForNewEntriesTask implements Runnable {
+
+        private final OpReadEntry op;
+        private final ReadEntriesCallback callback;
+        private final Object ctx;
+        private ScheduledFuture<?> scheduledFuture;
+        /**
+         * 0: init
+         * 1: running
+         * 2: cancelled
+         * 3: done
+         */
+        private final AtomicInteger state = new AtomicInteger();
+
+        public DelayCheckForNewEntriesTask(OpReadEntry op, ReadEntriesCallback callback, Object ctx) {
+            this.op = op;
+            this.callback = callback;
+            this.ctx = ctx;
+            start();
+        }
+
+        private void start() {
+            scheduledFuture = ledger.getScheduledExecutor().schedule(this,
+                    getConfig().getNewEntriesCheckDelayInMillis(), TimeUnit.MILLISECONDS);
+        }
+
+        @Override
+        public void run() {
+            if (!state.compareAndSet(0, 1)) {
+                return;
+            }
+            synchronized (pendingReadOpMutex) {
```

Comment:
Actually I think making `state` atomic is unnecessarily complicated.

All modifications on `state` are:
- CAS(0, 1) in `run` without `synchronized (pendingReadOpMutex)`
- CAS(1, 3) in `run` with `synchronized (pendingReadOpMutex)`
- CAS(0, 2) in `cancel`, which is always called in `cancelPendingReadRequest`'s `synchronized (pendingReadOpMutex)` code block

The only read operation of `state` is `isDone()`, which is always called in `cancelPendingReadRequest`'s `synchronized (pendingReadOpMutex)` code block.

i.e. only CAS(0, 1) in `run()` is not protected by `pendingReadOpMutex`.

Reply:
Addressed